### PR TITLE
Vim movement

### PR DESCRIPTION
**Pull request by robber-m**
Add some vim screen movement bindings

### CODE REVIEW

**Comment by ryabouza** [comment]
What a great pull request. 10/10 would review again.

**Comment by robber-m** [comment]
oops. maybe I need to start using different branches

**Comment by robber-m** [comment]
@ryabouza fortunately you have that privilege!

File: ce.c [hunk 1]
```diff
@@ -331,6 +438,11 @@ bool ce_move_cursor_to_soft_beginning_of_line(Buffer* buffer, Point* cursor)
      return true;
 }
 
+// underscores are not treated as punctuation for vim movement
+bool ce_ispunct(int c){
+     return c != '_' && ispunct(c);
```
Comment (justy989) [inline ce.c:614]: lol

File: .gitignore [hunk 2]
```diff
@@ -1,9 +1,34 @@
 # Object files
 *.o
+*.ko
```
Comment (justy989) [inline .gitignore:3]: boo !

File: .gitignore [hunk 3]
```diff
@@ -1,9 +1,34 @@
...
+*.obj
```
Comment (justy989) [inline .gitignore:4]: boo !

File: .gitignore [hunk 4]
```diff
@@ -1,9 +1,34 @@
...
+*.elf
```
Comment (justy989) [inline .gitignore:5]: boo !

File: .gitignore [hunk 5]
```diff
@@ -1,9 +1,34 @@
...
+
+# Precompiled Headers
+*.gch
```
Comment (justy989) [inline .gitignore:8]: boo !

File: .gitignore [hunk 6]
```diff
@@ -1,9 +1,34 @@
...
+*.pch
```
Comment (justy989) [inline .gitignore:9]: boo !

File: .gitignore [hunk 7]
```diff
@@ -1,9 +1,34 @@
...
+
+# Libraries
+*.lib
```
Comment (justy989) [inline .gitignore:12]: boo !

File: .gitignore [hunk 8]
```diff
@@ -1,9 +1,34 @@
...
+*.a
```
Comment (justy989) [inline .gitignore:13]: boo !

File: .gitignore [hunk 9]
```diff
@@ -1,9 +1,34 @@
...
+*.la
```
Comment (justy989) [inline .gitignore:14]: boo !

File: .gitignore [hunk 10]
```diff
@@ -1,9 +1,34 @@
...
+*.lo
```
Comment (justy989) [inline .gitignore:15]: boo !

File: .gitignore [hunk 11]
```diff
@@ -1,9 +1,34 @@
...
 
 # Shared objects (inc. Windows DLLs)
+*.dll
```
Comment (justy989) [inline .gitignore:18]: boo !

File: .gitignore [hunk 12]
```diff
@@ -1,9 +1,34 @@
...
 *.so
+*.so.*
```
Comment (justy989) [inline .gitignore:20]: boo !

File: .gitignore [hunk 13]
```diff
@@ -1,9 +1,34 @@
...
+*.dylib
```
Comment (justy989) [inline .gitignore:21]: boo !

File: .gitignore [hunk 14]
```diff
@@ -1,9 +1,34 @@
...
 
 # Executables
 ce
+*.exe
```
Comment (justy989) [inline .gitignore:25]: boo !

File: .gitignore [hunk 15]
```diff
@@ -1,9 +1,34 @@
...
+*.out
```
Comment (justy989) [inline .gitignore:26]: boo !

File: .gitignore [hunk 16]
```diff
@@ -1,9 +1,34 @@
...
+*.app
```
Comment (justy989) [inline .gitignore:27]: boo !

File: .gitignore [hunk 17]
```diff
@@ -1,9 +1,34 @@
...
+*.i*86
```
Comment (justy989) [inline .gitignore:28]: boo !

File: .gitignore [hunk 18]
```diff
@@ -1,9 +1,34 @@
...
+*.x86_64
```
Comment (justy989) [inline .gitignore:29]: boo !

File: .gitignore [hunk 19]
```diff
@@ -1,9 +1,34 @@
...
+*.hex
```
Comment (justy989) [inline .gitignore:30]: boo !

File: .gitignore [hunk 20]
```diff
@@ -1,9 +1,34 @@
...
 
+# Debug files
+*.dSYM/
```
Comment (justy989) [inline .gitignore:33]: boo !

File: .gitignore [hunk 21]
```diff
@@ -1,9 +1,34 @@
...
+*.su
```
Comment (justy989) [inline .gitignore:34]: boo !

File: ce.c [hunk 22]
```diff
@@ -487,6 +489,113 @@ int64_t ce_find_delta_to_char_backward_in_line(Buffer* buffer, const Point* loca
...
+                         delta->y = line - location->y;
+                         return true;
+                    }
+               }
+               else if(*iter_char == *cur_char) counter++;
+               iter_char += d;
+          }
+
+          do i++;
+          while(!(line_str = buffer->lines[line += d]) && i < n_lines);
```
Comment (justy989) [inline ce.c:565]: woah I've never seen a do while loop without braces before !

File: ce.c [hunk 23]
```diff
@@ -622,20 +758,41 @@ bool ce_insert_newline(Buffer* buffer, int64_t line)
...
+          ce_message("%s() specified line %lld ouside of buffer, which has %lld lines", __FUNCTION__, line, buffer->line_count);
+          return false;
+     }
+
+     if(line == buffer->line_count - 1) return true; // nothing to do
+     char* l1 = buffer->lines[line];
+     size_t l1_len = strlen(l1);
+     char* l2 = buffer->lines[line+1];
+     size_t l2_len = strlen(l2);
+     buffer->lines[line] = l1 = realloc(l1, l1_len + l2_len + 2); //space and null
```
Comment (justy989) [inline ce.c]: not a fan of multiple assignments in one line !

File: ce.c [hunk 24]
```diff
@@ -909,7 +1066,44 @@ int64_t ce_find_delta_to_end_of_line(const Buffer* buffer, Point* cursor)
...
+
+bool ce_set_cursor(const Buffer* buffer, Point* cursor, const Point* location)
+{
+     CE_CHECK_PTR_ARG(buffer);
+     CE_CHECK_PTR_ARG(cursor);
+     CE_CHECK_PTR_ARG(location);
+
+     Point dst = *location;
+
+     if(dst.x < 0) dst.x = 0;
```
Comment (justy989) [inline ce.c:1082]: looks like ce_set_cursor and ce_move_cursor share a lot of the same logic. Should we make ce_move_cursor call ce_set_cursor?

File: ce.c [hunk 25]
```diff
@@ -1015,13 +1209,13 @@ bool ce_follow_cursor(const Point* cursor, int64_t* left_collumn, int64_t* top_r
           *top_row = bottom_row - view_height;
      }
 
-     int64_t right_collumn = *left_collumn + view_width;
+     int64_t right_column = *left_column + view_width;
```
Comment (justy989) [inline ce.c:1213]: neccesary is another word I commonly misspell !

File: ce_config.c [hunk 26]
```diff
@@ -161,7 +194,7 @@ BufferNode* new_buffer_from_file(BufferNode* head, const char* filename)
           return NULL;
      }
 
-     if(!initialize_buffer(buffer)){
+     if(!initialize_buffer(buffer)){
```
Comment (justy989) [inline ce_config.c:197]: boo !

P.S. Ryan is the best! 👯

File: ce_config.c [hunk 27]
```diff
@@ -788,6 +1066,148 @@ bool key_handler(int key, BufferNode* head, void* user_data)
...
+                         dup(out_fds[1]); // new stdout
+                         int64_t cursor_position = cursor->x+1;
+                         for(int64_t y_itr = 0; y_itr < cursor->y; y_itr++){
+                              cursor_position += strlen(buffer->lines[y_itr]);
+                         }
+                         char* cursor_arg;
+                         asprintf(&cursor_arg, "-cursor=%"PRId64, cursor_position);
+
+                         char* line_arg;
+                         asprintf(&line_arg, "-lines=%"PRId64":%"PRId64, begin_format_line+1, end_format_line+1);
```
Comment (justy989) [inline ce_config.c:1112]: Are these lines something we have to free? Even though we exit right below, can the os reclaim memory on forked processes? Or should we just free it so we can have clean valgrind runs?

File: ce_config.c (hunk 26 above)
Comment (justy989) [reply ce_config.c:197]: I can't believe I said that !

File: ce_config.c (hunk 27 above)
Comment (robber-m) [reply ce_config.c:1112]: Done!

File: ce.c (hunk 23 above)
Comment (robber-m) [reply ce.c]: Changed

**Review by justy989 — APPROVED** [review]

**Pull request merged**